employee-rostering-gwtui/src/main/java/org/optaweb/employeerostering/gwtui/client/pages/rotation/ShiftTemplateEditForm.java

```diff
@@ -0,0 +1,262 @@
+    protected void onClose() {
+        if (shiftTemplateGridObject != null) {
+            shiftTemplateGridObject.setSelected(false);
+        }
+    }
+
+    @EventHandler("apply-button")
+    public void onApplyButtonClick(@ForEvent("click") final MouseEvent e) {
+        if (shiftTemplateGridObject != null) {
+
+            final ShiftTemplateView template = shiftTemplateGridObject.getShiftTemplateModel().getShiftTemplateView();
+            final Long oldEmployee = template.getRotationEmployeeId();
+            final Duration oldDurationFromReference = template.getDurationBetweenRotationStartAndTemplateStart();
+            final Duration oldDurationOfTemplate = template.getShiftTemplateDuration();
+
+            if (updateShiftTemplateFromWidgetsIfValid(template)) {
+                RotationRestServiceBuilder.updateShiftTemplate(template.getTenantId(), template, FailureShownRestCallback.onSuccess((final ShiftTemplateView updatedShiftTemplate) -> {
+                    shiftTemplateGridObject.getShiftTemplateModel().withShiftTemplateView(updatedShiftTemplate);
+                    shiftTemplateGridObject.setSelected(false);
+                    eventManager.fireEvent(ROTATION_INVALIDATE);
+                    hide();
+                }).onFailure(i -> {
+                    template.setRotationEmployeeId(oldEmployee);
+                    template.setDurationBetweenRotationStartAndTemplateStart(oldDurationFromReference);
+                    template.setShiftTemplateDuration(oldDurationOfTemplate);
+                }).onError(i -> {
+                    template.setRotationEmployeeId(oldEmployee);
+                    template.setDurationBetweenRotationStartAndTemplateStart(oldDurationFromReference);
+                    template.setShiftTemplateDuration(oldDurationOfTemplate);
+                }));
+            }
+        } else {
+            final ShiftTemplateView template = new ShiftTemplateView();
+            template.setTenantId(tenantStore.getCurrentTenantId());
+
+            if (updateShiftTemplateFromWidgetsIfValid(template)) {
+                template.setRotationEmployeeId(parseId(employeeSelect.getSelectedValue()));
+                Duration newDurationFromReference = Duration.ofDays(from.getDayOffset()).plusSeconds(from.getTime().toSecondOfDay());
+                Duration newDurationOfTemplate;
+                if (to.getDayOffset() >= from.getDayOffset()) {
+                    newDurationOfTemplate = Duration.ofDays(to.getDayOffset()).plusSeconds(to.getTime().toSecondOfDay()).minus(newDurationFromReference);
+                } else {
+                    newDurationOfTemplate = Duration.ofDays(to.getRotationLength()).minus(newDurationFromReference).plusDays(to.getDayOffset()).plusSeconds(to.getTime().toSecondOfDay());
+                }
+                template.setDurationBetweenRotationStartAndTemplateStart(newDurationFromReference);
+                template.setShiftTemplateDuration(newDurationOfTemplate);
```

Comment:
This (lines 194-203) is still duplicate of the body of `updateShiftTemplateFromWidgetsIfValid()`.

Reply:
Fixed